modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/PeriodicExpiryMessageDeletionTask.java

```diff
@@ -0,0 +1,180 @@
+ */
+
+package org.wso2.andes.kernel;
+
+import com.google.common.util.concurrent.SettableFuture;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.kernel.slot.AbstractSlotManager;
+import org.wso2.andes.kernel.slot.SlotManagerClusterMode;
+import org.wso2.andes.kernel.slot.SlotManagerStandalone;
+import org.wso2.andes.store.FailureObservingStoreManager;
+import org.wso2.andes.store.HealthAwareStore;
+import org.wso2.andes.store.StoreHealthListener;
+import org.wso2.andes.tools.utils.MessageTracer;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * PeriodicExpiryMessageDeletionTask is responsible to delete the expired messages from the database which were not currently
+ * allocated to any slots. If we delete any messages that are assigned to a slot delivery worker, the
+ * content may got deleted when the content is asked to retrieve. This can mess up the message delivery
+ * flow. Because of that the messages not allocated to a slot are considered as safe to delete
+ */
+public class PeriodicExpiryMessageDeletionTask implements Runnable, StoreHealthListener {
+
+    private static Log log = LogFactory.getLog(PeriodicExpiryMessageDeletionTask.class);
+
+    /**
+     * Indicates and provides a barrier if messages stores become offline.
+     * marked as volatile since this value could be set from a different thread
+     */
+    protected volatile SettableFuture<Boolean> messageStoresUnavailable;
+
+    /**
+     * Used to perform database operations on the context store.
+     */
+    private AndesContextStore andesContextStore = AndesContext.getInstance().getAndesContextStore();
+
+    /**
+     * Holds the slot manager based on broker running mode
+     */
+    private AbstractSlotManager abstractSlotManagerSlotManager;
+
+    /**
+     * Indicate the cluster mode is enabled or not
+     */
+    private boolean isClusteringEnabled;
+
+    public PeriodicExpiryMessageDeletionTask() {
+
+        this.messageStoresUnavailable = null;
+        // Register AndesRecoveryTask class as a StoreHealthListener
+        FailureObservingStoreManager.registerStoreHealthListener(this);
+
+        isClusteringEnabled = AndesContext.getInstance().isClusteringEnabled();
+        // Set the appropriate slot manager
+        if (isClusteringEnabled) {
+            this.abstractSlotManagerSlotManager = SlotManagerClusterMode.getInstance();
+        } else {
+            this.abstractSlotManagerSlotManager = SlotManagerStandalone.getInstance();
+        }
+
+    }
+
+    /**
+     * Get the expiry messages queue wise from unallocated region  and delete those from DB
+     */
+    protected void deleteExpiredMessages(){
+
+        try {
+            /**
+             * This logic belongs to an MB run in stand alone mode / the coordinator node run in cluster mode
+             */
+            if (!isClusteringEnabled
```

Comment:
what do u think about this?

`class PeriodicExpiryMessageDeletionTask implements Runnable {

	interface DeletionTask {

		void delete();

	}

	DeletionTask task;

	class CoordinatorOnlyDeletionTask implements DeletionTask {

		@Override
		public void delete() {

			if (AndesContext.getInstance().getClusterAgent().isCoordinator()) {
				deletestuff();
			}

		}

	}

	class StandaloneDeletionTask implements DeletionTask {

		@Override
		public void delete() {
			deletestuff();
		}

	}

	public void deletestuff() {

	}

	public PeriodicExpiryMessageDeletionTask() {

		if (isClusteringEnabled) {
			task = new CoordinatorOnlyDeletionTask();
		} else {
			task = new StandaloneDeletionTask();
		}

	}

	public void run() {
		task.delete();
	}
}
`